src/Statistics.jl

```diff
@@ -0,0 +1,115 @@
+
+Finally `stop_iter` denotes the maximal number of iterations to perform and
+the `kwargs...` are passed to [`isapprox`](@ref) to stop, when the inimal change
+between two iterates is small. For more stopping criteria
+check the [`Manopt.jl`](https://manoptjl.org) package and use a solver therefrom.
+
+The algorithm is further described in 
+> Afsari, B; Tron, R.; Vidal, R.: On the Convergence of Gradient
+> Descent for Finding the Riemannian Center of Mass,
+> SIAM Journal on Control and Optimization (2013), 51(3), pp. 2230–2260, 
+> doi: [10.1137/12086282X](https://doi.org/10.1137/12086282X),
+> arxiv: [1201.0925](https://arxiv.org/abs/1201.0925">1201.0925)
+"""
+function mean(M::Manifold, x::AbstractVector; kwargs...)
+    y = x[1]
+    return mean!(M, y, x; kwargs...)
+end
+mean(M::Euclidean, x, w) = mean(x,w)
+function mean!(M::Manifold, y, x::AbstractVector;
+            weights = ones(length(x)) / length(x),
+            stop_iter=100,
+            kwargs...
+        ) where {T}
+    yold = y
+    v = zero_tangent_vector.(Ref(M), fill(y,length(x)))
+    for i=1:stop_iter
+        yold = y
+        log!.(Ref(M), v, Ref(yold), x)
+        y = exp(M, yold, sum( weights.*v ) / 2 )
+        isapprox(M,y,yold; kwargs...) && break
+    end
+    return y
+end
+
+@doc doc"""
+    median(M,x; weights=1/n*ones(n), stop_tol=10^-10, stop_iter=10000, use_rand = false )
+
+computes the Riemannian median of the vector `x`  of `n` points on the
+[`Manifold`](@ref) `M`. This function is nonsmooth (i.e nondifferentiable) and
+uses a cyclic proximal point scheme. Optionally one can provide
+weights $w_i$ for the weighted Riemannian median. The general formula to compute
+the minimizer reads
+````math
+\argmin_{y\in\mathcal M}\sum_{i=1}^n w_i\mathrm{d}_{\mathcal M}(y,x_i),
+````
+where $\mathrm{d}_{\mathcal M}$ denotes the Riemannian [`distance`](@ref).
+
+The cyclic proximal point can be run with a random cyclic order by setting
+`use_rand` to `true`
+
+Finally `stop_iter` denotes the maximal number of iterations to perform and
+the `kwargs...` are passed to [`isapprox`](@ref) to stop, when the inimal change
+between two iterates is small. For more stopping criteria
+check the [`Manopt.jl`](https://manoptjl.org) package and use a solver therefrom.
+
+The algorithm is further described in Algorithm 4.3 and 4.4 in 
+> Bačák, M: Computing Medians and Means in Hadamard Spaces.
+> SIAM Journal on Optimization (2014), 24(3), pp. 1542–1566,
+> doi: [10.1137/140953393](https://doi.org/10.1137/140953393),
+> arxiv: [1210.2145](https://arxiv.org/abs/1210.2145)
+"""
+function median(M::Manifold, x::AbstractVector; kwargs...)
+    y = x[1]
+    return median!(M, y, x; kwargs...)
+end
+function median!(M::Manifold, y, x::AbstractVector;
+            weights= ones(length(x)) / length(x),
+            stop_iter=10000,
+            use_random = false,
+            kwargs...
+        ) where {T}
+    n = length(x)
+    yold = y
```

Comment:
We can't really assume that `y` is a valid starting point. `copy(x[1])` should be, or perhaps we could specify the starting point in a keyword argument?

Reply:
Until now for all input we always assumed everything is valid (for exp and log for example) and fine and only checked for `ArrayManifold` that points actually are on the manifold (one can have a lot of fun with non-normed points on the sphere giving strange results). So I would opt to stay with that scheme here, too.

Reply:
But `y` is output and not input? If that's the starting point it should be in the docs (and non-mutating mean/median shouldn't use `similar_result`).

Reply:
So what do you propose to do? I am fine with documenting that the output y is also used as starting point

Reply:
Right, I think using `y` as a starting point is the most reasonable option. The non-mutating variant should then supply a copy of the first point as `y` here then.

Reply:
Tha's actually now missing, that `similar_result` works, it filly `y` basically with zeros instead of `x[1]`.